### PR TITLE
RFD 0073: Public Image Registry

### PR DESCRIPTION
**Pull request by logand22**
This PR includes an RFD on migrating from Quay.io to Amazon Public ECR. See https://github.com/gravitational/teleport/blob/rfd/0073-public-image-registry/rfd/0073-public-image-registry.md for a rendered version.

I'd like to gather feedback around the use of Amazon Public ECR over the alternatives to make sure we understand the trade offs before making a final decision.

Note: We are in the process of leveraging Amazon ECR for Cloud images. See https://github.com/gravitational/cloud/issues/1162 for more information. 

It's important to also note that Amazon ECR and Amazon ECR Public are two separate products offered by Amazon with different features. See [Amazon ECR](https://docs.aws.amazon.com/ecr/index.html) for more information. 

Related Issue:
* https://github.com/gravitational/teleport/issues/5225

### CODE REVIEW

File: rfd/0073-public-image-registry.md
Comment (wadells) [inline rfd/0073-public-image-registry.md:31]: Mirroring to quay.io cannot be out of scope, since we must support extant deployments in the field.  I agree that mirroring to docker hub or other repos can be left out of scope for now.

File: rfd/0073-public-image-registry.md
Comment (wadells) [inline rfd/0073-public-image-registry.md:45]: Can you add where quay.io fits in this diagram?

File: rfd/0073-public-image-registry.md
Comment (wadells) [inline rfd/0073-public-image-registry.md:49]: I'd like to see a discussion of which users will have access to read/write from which repos.  I recognize this rehashes Artifact Storage Standards to a degree, but I believe it is worth restating in order to have this RFD stand on its own.

File: rfd/0073-public-image-registry.md
Comment (wadells) [inline rfd/0073-public-image-registry.md:54]: "changes to images" >>  Can you discuss tag immutability as it does (or doesn't) apply to the public ecr repos?  I thought to some extent we'd be preventing changes to images in line with our "once a tag is published, it never changes" policy.

File: rfd/0073-public-image-registry.md
Comment (wadells) [inline rfd/0073-public-image-registry.md:99]: `teleport` or `gravitational`?  The diagram above says `gravitational`

We should probably squat both, even if we only use one.

**Review by wadells — COMMENTED** [review]
Thanks for putting this together!

First round of thoughts attached.

File: rfd/0073-public-image-registry.md
Comment (logand22) [reply rfd/0073-public-image-registry.md:167]: You are right. I don't believe they need to be mirrored. They are apart of the [build-package.sh](https://github.com/gravitational/teleport/blob/99ad5c59a4195bc450dabfb4f7f377d90ccda8e6/build.assets/build-package.sh). As its internal I don't think external contributors reference it directly.

File: rfd/0073-public-image-registry.md
Comment (logand22) [reply rfd/0073-public-image-registry.md:99]: The public repository would already be using `gravitational` as it's registry alias, therefore `teleport` in this case refers to public.ecr.aws/gravitational/*teleport*. 

I'm open to having `gravitational`, `teleport` or both as the registry alias for public ECR. I would note the recent terraform provider went with `gravitational/teleport`.

File: rfd/0073-public-image-registry.md
Comment (logand22) [reply rfd/0073-public-image-registry.md]: Changed to be more explicit in https://github.com/gravitational/teleport/pull/13165/commits/6378fea6c5dbbae3b19033dd515e0dd0aab26409

File: rfd/0073-public-image-registry.md
Comment (logand22) [reply rfd/0073-public-image-registry.md]: Changed to be more explicit in https://github.com/gravitational/teleport/pull/13165/commits/6378fea6c5dbbae3b19033dd515e0dd0aab26409

File: rfd/0073-public-image-registry.md
Comment (logand22) [reply rfd/0073-public-image-registry.md]: Changed to be more explicit in https://github.com/gravitational/teleport/pull/13165/commits/6378fea6c5dbbae3b19033dd515e0dd0aab26409

File: rfd/0073-public-image-registry.md
Comment (logand22) [reply rfd/0073-public-image-registry.md]: Updated in https://github.com/gravitational/teleport/pull/13165/commits/6378fea6c5dbbae3b19033dd515e0dd0aab26409

File: rfd/0073-public-image-registry.md
Comment (logand22) [reply rfd/0073-public-image-registry.md]: Changed to be more explicit in https://github.com/gravitational/teleport/pull/13165/commits/6378fea6c5dbbae3b19033dd515e0dd0aab26409

File: rfd/0073-public-image-registry.md
Comment (logand22) [reply rfd/0073-public-image-registry.md]: Removed these keys to be more explicit in https://github.com/gravitational/teleport/pull/13165/commits/6378fea6c5dbbae3b19033dd515e0dd0aab26409

File: rfd/0073-public-image-registry.md
Comment (logand22) [reply rfd/0073-public-image-registry.md]: Good point, added an exception here: https://github.com/gravitational/teleport/pull/13165/commits/e5cd72e93f513f3a71c4f32bb9f75f7844efc650

File: rfd/0073-public-image-registry.md
Comment (logand22) [reply rfd/0073-public-image-registry.md:55]: Add a short discussion on the main types of users with permissions: service accounts for CI/CD, teleport employees and release engineers in https://github.com/gravitational/teleport/pull/13165/commits/5fe26fa323ddbfa2daabfd84d93a85c077c4a0a6. 

Should I go into more specifics?

File: rfd/0073-public-image-registry.md
Comment (logand22) [reply rfd/0073-public-image-registry.md]: Amazon ECR Public doesn't support tag immutability. This will be enforced in the first stage with Amazon ECR. You are correct that we would strive to have a "once a tag is published, it never changes" policy. However, RFD 0017 for artifact storage standards don't specify nor enforce that policy. It only states that administrative changes must be through an assumed role and are audited w/ alerts for any such modifications. 

For instance, an alert should be set up such that any write to the public ECR repository that doesn't come from the service account user should trigger an alert. It doesn't state that this can't happen, only that the security team should be alerted. 

I'm open to the idea of limiting even admin users from making changes even if they are audited. Would be cool if a role was available to make changes but was only ever given via short-lived credentials that are granted after being approved via a PR.

File: rfd/0073-public-image-registry.md
Comment (jakule) [inline rfd/0073-public-image-registry.md:15]: What about GCP? We're already using it for some of our images and GCP offers images scanning (and also lacks of tag immutability AFAIK)

Comment (logand22) [reply rfd/0073-public-image-registry.md:15]: [GCR](https://cloud.google.com/container-registry) and [Artifact Registry](https://cloud.google.com/artifact-registry) are both decent solutions for a container registry. From what I understand, they support vulnerability scanning with a cost associated, albeit small ($0.26 per image scan), whereas AWS ECR has a free tier. 

There doesn't seem to be a specific feature difference between the two. We already host cloud images on ECR as well as our other artifacts such as terraform provider and linux packages. Is there any specific reason we should leverage GCP over AWS in this case?

File: rfd/0073-public-image-registry.md
Comment (wadells) [reply rfd/0073-public-image-registry.md:57]: I think https://github.com/gravitational/teleport/commit/5fe26fa323ddbfa2daabfd84d93a85c077c4a0a6 is plenty of context. Thanks!

File: rfd/0073-public-image-registry.md
Comment (wadells) [reply rfd/0073-public-image-registry.md]: > Amazon ECR Public doesn't support tag immutability. 

Gotcha.  I think this was the detail I was missing.  You might explicitly call this out as a limitation of Amazon ECR Public.

File: rfd/0073-public-image-registry.md
Comment (wadells) [reply rfd/0073-public-image-registry.md:112]: Whoops -- my bad.  I was mistaking the repository for the org (or "registry alias"?).  Where is the org/registry alias defined?  I don't see that in this example TF.

**Review by wadells — APPROVED** [review]
LGTM!

```
      ___           ___                       ___                                ___     
     /\  \         /\__\          ___        /\  \                   ___        /\  \    
    /::\  \       /:/  /         /\  \      /::\  \                 /\  \       \:\  \   
   /:/\ \  \     /:/__/          \:\  \    /:/\:\  \                \:\  \       \:\  \  
  _\:\~\ \  \   /::\  \ ___      /::\__\  /::\~\:\  \               /::\__\      /::\  \ 
 /\ \:\ \ \__\ /:/\:\  /\__\  __/:/\/__/ /:/\:\ \:\__\           __/:/\/__/     /:/\:\__\
 \:\ \:\ \/__/ \/__\:\/:/  / /\/:/  /    \/__\:\/:/  /          /\/:/  /       /:/  \/__/
  \:\ \:\__\        \::/  /  \::/__/          \::/  /           \::/__/       /:/  /     
   \:\/:/  /        /:/  /    \:\__\           \/__/             \:\__\       \/__/      
    \::/  /        /:/  /      \/__/                              \/__/                  
     \/__/         \/__/      
```

**Review by jakule — APPROVED** [review]
LGTM

File: rfd/0073-public-image-registry.md
Comment (logand22) [reply rfd/0073-public-image-registry.md:112]: I don't have permissions currently. I will start on the `core-release-engineer-admin` role since regardless of what we end up doing that will be useful.

File: rfd/0073-public-image-registry.md
Comment (marcoandredinis) [inline rfd/0073-public-image-registry.md:172]: We are creating a Kubernetes Operator for Teleport
It'll have its own image: `teleport-operator`

I understand this document only reflects what we currently have, but can we add a note somewhere so we don't forget?
We expect to release the operator in the next couple of weeks

**Review by marcoandredinis — APPROVED** [review]
Just left some comments, but mostly :nail_care:

File: rfd/0073-public-image-registry.md
Comment (logand22) [reply rfd/0073-public-image-registry.md:73]: @marcoandredinis This Cloud RFD on [Artifact Storage Standards](https://github.com/gravitational/cloud/blob/9124947fdfb0773fa9bd567160481bed4ec84b7e/rfd/0017-artifact-storage-standards.md#levels) defines the breakdown and criteria for the levels. 

In this case teleport, teleport-ent, teleport-plugin-*, and the new image teleport-operator would all be considered in the gold level as they directly secure our customers infrastructure. 

Do you think I should elevate some of that discussion into this RFD to make it more clear how these images should be categorized?

**Review by jimbishopp — APPROVED** [review]

File: rfd/0073-public-image-registry.md
Comment (marcoandredinis) [reply rfd/0073-public-image-registry.md:73]: I missed that relation between this RFD and the other RFD described in the first item, my bad

**Review by marcoandredinis — APPROVED** [review]

**Review by tcsc — APPROVED** [review]

File: rfd/0073-public-image-registry.md
Comment (xinding33) [reply rfd/0073-public-image-registry.md:112]: Commenting here because @logand22 reached out to me asking if I had opinions on rebranding images from `public.ecr.aws/gravitational/teleport` to `public.ecr.aws/teleport/teleport`, thereby dropping the `gravitational` brand from container images. Personally, I find that a complete rebrand across the board is not at the top of the priority list so imo, going with `/gravitational/teleport` is preferable for consistency (mainly less confusing for users). I agree that we should squat `teleport` too if/when we decide to completely swap everything over.

File: rfd/0073-public-image-registry.md
Comment (r0mant) [inline rfd/0073-public-image-registry.md:245]: Aren't Gravity customers still using these?

File: rfd/0073-public-image-registry.md
Comment (r0mant) [inline rfd/0073-public-image-registry.md:60]: This is good but there should be some manual escape hatch allowing us to publish images manually in case of an urgent fix or something which we sometimes need to do. If it's via "release engineer" credentials it's fine, let's just make sure that this process is documented somewhere.

**Review by r0mant — APPROVED** [review]

**Review by gozer — APPROVED** [review]

File: rfd/0073-public-image-registry.md
Comment (logand22) [reply rfd/0073-public-image-registry.md:245]: I'll have to defer to @wadells on that. I do know that they haven't been updated in Quay in >2 years and don't seem to have any activity as far as Quay reports.

File: rfd/0073-public-image-registry.md
Comment (logand22) [reply rfd/0073-public-image-registry.md:60]: There is a `release-engineer-core-admin` role that is created that will have capabilities to do manual fixes in case of an urgent fix. I'll add more information on that to clarify.

Comment (logand22) [reply rfd/0073-public-image-registry.md:60]: Added in https://github.com/gravitational/teleport/pull/13165/commits/2c4eb54ee73c19b03169fc596bc0a65ec0eb3111

File: rfd/0073-public-image-registry.md
Comment (wadells) [reply rfd/0073-public-image-registry.md:247]: Hm...

As @logand22  mentioned, both of these repos have a single tag pushed ~2 years ago, and more importantly: No pulls on records for the past 3 months, during which we published several gravity releases and done many more installs.  Compare this to more actively used gravity images which show pulls:

https://quay.io/repository/gravitational/rig?tab=logs
https://quay.io/repository/gravitational/satellite?tab=logs

As far as my knowledge of the build system goes, monitoring app images are baked & distributed in a couple different ways, none of which depends on quay for outgoing images that customers will download:

1.  During our builds, pre-built copies are pulled from s3://clientbuilds.gravitational.io/monitoring-app/
2. The aforementioned bucket is populated via the gravity build process, which (if it can't download a tarball) will clone the monitoring app source, build and upload
3. For enterprise releases, base images (including monitoring app) are published to get.gravitational.io where customers can download them
4. For OSS releases, base images (including monitoring app) are published to s3://hub.gravitational.io

Best as I can tell, these aren't upstreams to monitoring-app either.  They're mentioned here:

https://github.com/gravitational/monitoring-app/blob/d596a3d36f910eaf69abda42832da9e3e21e5b0c/images/Makefile#L34-L37

However anything that would be pushed/pulled from that makefile can also be derived from scratch from the Docker file. 

That said, there is no harm in moving `monitoring-grafana` and `monitoring-influxdb` to the "don't migrate/don't maintain list".  We can always delete them later when we have no contractual obligations in regards to Gravity.

File: rfd/0073-public-image-registry.md
Comment (webvictim) [reply rfd/0073-public-image-registry.md]: Finally got around to reading this. They don't need to be publicly mirrored (unless external contributors are running `build-package.sh` themselves - I just mirrored them when we started using them to shield us from the repo/tag randomly disappearing and to harden the security chain.

**Pull request merged**
